contrib/IECoreUSD/src/IECoreUSD/PrimitiveAlgo.cpp

```diff
@@ -239,13 +245,111 @@ void readPrimitiveVariable( const pxr::UsdGeomPrimvar &primVar, pxr::UsdTimeCode
 	primitive->variables[name] = IECoreScene::PrimitiveVariable( interpolation, data, indices );
 }
 
+pxr::UsdSkelCache *skelCache()
+{
+	// the SkelCache is thead-safe and is documented as intended to persist
+	// and be used across stages, so we use a global cache for the process.
+	static pxr::UsdSkelCache *g_skelCache = nullptr;
+	if( !g_skelCache )
+	{
+		g_skelCache = new pxr::UsdSkelCache;
+	}
+
+	return g_skelCache;
+}
+
+bool readPrimitiveVariables( const pxr::UsdSkelRoot &skelRoot, const pxr::UsdGeomPointBased &pointBased, pxr::UsdTimeCode time, IECoreScene::Primitive *primitive )
+{
+	pxr::UsdSkelSkeletonQuery skelQuery = ::skelCache()->GetSkelQuery( pxr::UsdSkelBindingAPI( pointBased.GetPrim() ).GetInheritedSkeleton() );
+	if( !skelQuery )
+	{
+		return false;
+	}
+
+	pxr::VtMatrix4dArray skinningXforms;
+	if( !skelQuery.ComputeSkinningTransforms( &skinningXforms, time ) )
+	{
+		return false;
+	}
+
+#if USD_VERSION < 2011
+	::skelCache()->Populate( skelRoot );
+#else
+	::skelCache()->Populate( skelRoot, pxr::UsdTraverseInstanceProxies() );
+#endif
```

Comment:
Each method on `UsdSkelCache` states whether calling `Populate` is required (unhelpfully on a secondary sentance of the doc string, so you have to scroll to the details section of the docs to see it). `GetSkinningQuery` is the only method that requires calling `Populate` first.

Reply:
Sounds good! Wish I'd read that first as it would have saved me a bit of wandering around the USD source not really understanding what I was looking at :)